compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotShrinker.kt

```diff
@@ -105,15 +104,26 @@ object ClasspathSnapshotShrinker {
             // No need to collect supertypes outside the given set of classes (e.g., "java/lang/Object")
             @Suppress("SimpleRedundantLet")
             classIdToClassSnapshot[classId]?.let {
-                it.classSnapshot.getSupertypes(classNameToClassIdResolver).filter { supertype -> supertype in classIds }.toSet()
+                it.getSupertypes(classNameToClassIdResolver).filterTo(mutableSetOf()) { supertype -> supertype in classIds }
             } ?: emptySet()
         }
 
-        val referencedClassIds = referencedClasses.map { it.classSnapshot.getClassId() }.toSet()
+        val referencedClassIds = referencedClasses.mapTo(mutableSetOf()) { it.classId }
         val transitivelyReferencedClassIds: Set<ClassId> =
             ImpactAnalysis.findImpactedClassesInclusive(referencedClassIds, supertypesResolver) // Use Set for presence check
 
-        return allClasses.filter { it.classSnapshot.getClassId() in transitivelyReferencedClassIds }
+        return allClasses.filter { it.classId in transitivelyReferencedClassIds }
+    }
+
+    class MetricsReporter(
```

Comment:
Ivan's comment from https://github.com/google/kotlin/commit/38227b5843d51941f80b8b96585ef74fd363e860#r64567154

> "IMHO using the same pattern everywhere (BuildMetricsReporter directly) is better, for consistency."

Right, this pattern is not great. But I created this because this `ClasspathSnapshotShrinker` is called in 2 places (before and after the compilation), so the `BuildTime` enums can't be shared. So I need a way to force the caller to provide a list of `BuildTime`s based on the caller's context and pass it to the shrinker.